cores/nRF5/wiring_digital.c

```diff
@@ -18,27 +18,39 @@
 */
 
 #include "nrf.h"
-
 #include "Arduino.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+#define GPIO_PORT0 ((NRF_GPIO_Type*) 0x50000000UL)
```

Comment:
I have an idea that might simplify this, using the `digitalPinToPort` and `digitalPinToBitMask` macros in `/cores/nRF5/Arduino.h`. I'll make them and try to push to your branch so you can try them out.

Reply:
@jhmaloney I've pushed my changes in https://github.com/sandeepmistry/arduino-nRF5/pull/432/commits/652e38ab92eab4fcbfda547f08610dacec73b9fb and https://github.com/sandeepmistry/arduino-nRF5/pull/432/commits/3d82ada68832ed7b074f8d52501d54bc33409ae2 please try to pull and test if `pinMode`, 'digitalWrite`, `digitalRead` and `Wire` still work on your v2 board. I've tested on my v1 and v1.5 and things seem ok.

Reply:
Great! I see you've already made that change so I'll pull it and give it a try.

Reply:
I found a fixed two small bugs involving pins on port1. Also fixed a small bug in Wire_nRF52.cpp. With these fixes digitalWrite and the internal I2C bus work. I haven't yet tested digitalRead and the external I2C bus.

I've also made two small changes to the pin map for the Calliope mini: (1) added the ability to use connector pins 16 and 17 as analog inputs A6 and A7; and (2) swapped the positions of the microphone pin and an unassigned pin in the digital pin map. Since the microphone read as A0 and should never be used as a digital pin, change (2) should not impact any existing code. This change makes it easier to prevent digital access to the microphone pin (i.e. using a simple range check) since using that pin as a digital output could potentially damage the microphone amplifier (e.g. setting the pin to 0 would connect the output of the amplifier to ground).

Reply:
Have now tested digitalRead on pins from both ports and tested the external i2c bus. In the process, I found and fixed a bug in i2c handling of port1 pins).

I also found and fixed a long-standing bug in the pulseIn() function. (It should have been mapping the user level pin number to the internal pin number.)

Things are looking good!

Reply:
I fixed what I believe was a bug in pulseIn(). It should take a user pin number (e.g. the numbers used in digitalRead()) but some old (and incorrect) code to do that mapping was commented out. The computation of pulse microseconds from loops was also off by about 30%. I adjusted the constant based on tests against a range of pulse widths that were checked with an oscilloscope. Since pulse width measurements are quantized to an integral number of loops and subject to jitter due to interrupts, this function isn't expected to have microsecond precision, but I think the new calculation is much closer than the old one. (One use-case is decoding the signal from a DHT humidity/temperature sensor, which requires distinguishing between pulse widths of 28 and 70 usecs.)